### PR TITLE
Allow changing hair, facial hair, eye and skin colour

### PR DESCRIPTION
**Pull request by Saicchi**
#### Summary
Features "Allow non-debug change of hair, facial hair, eye colour and skin colour"

#### Purpose of change
Allow the player to change hair, facial hair, eye and skin colour without using debug mode.

#### Describe the solution

Add a new `Character` function to change appearance.
Add two new `iuse` functions to the hand mirror
Add a new `iexamine` function to the standing mirror.

#### Describe alternatives you've considered
None

#### Testing
- [x] Player can change eye colour
- [x] Player can cange hair
- [x] Player can change skin colour

### CODE REVIEW

**Comment by Entity-y** [comment]
Would there be a way to cycle through hair length over time as mentioned in #51112

**Comment by Theawesomeboophis** [comment]
Hair dye might be a good item to consider adding in the future as well.

**Comment by Saicchi** [comment]
Probably with this new effect on condition thing.
https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/EFFECT_ON_CONDITION.md

**Comment by Entity-y** [comment]
Maybe giving bleach an alternative use, making hair one shade lighter until eventually the hair is white

**Comment by Entity-y** [comment]
As far as changing eye color, this could be done artificially with contacts, without breaking immersion.

**Comment by Theawesomeboophis** [comment]
> Maybe giving bleach an alternative use, making hair one shade lighter until eventually the hair is white

That sounds like an interesting proposal, and maybe make it so it has a chance to fail, leading to your character accidentally consuming the bleach instead, I think that could be a really neat and humorous way to die.

**Comment by acepleiades** [comment]
As Erk once stated, currently the game allows you to change cosmetic vehicle shapes without cost, so cosmetic sprites could also be changed without cost. I recently worked on UltiCa character sprites such as skin tones and hairs and I suggested Sacchi to add a way to customize them as well, I'm really glad he got into it, I also mentioned this to Erk and he said he wouldn't be vehemently opposed to it. Although if you (plural) want to code an immersive way of doing these things I would also be in favor of that

**Comment by I-am-Erk** [comment]
I'm not vehemently opposed to this, but in general I would prefer not to do it this way. This is quite different from changing declarative elements like your name and gender. At the very least I'd prefer to have some discussion about alternate, more immersive ways to do it.

**Comment by acepleiades** [comment]
Someone's preferred pronouns are declarative but not their physical appearance (or at least not instantly), however, switching genders in game does also change the physical appearance to the degree that there is a difference between male and female sprites. In 32x32+ tilesets, this includes giving and removing breasts, something that would realistically require surgeries. By inference I would also like to be able to change other cosmetic sprites, and I could think of various roleplay reasons that would result in even me wanting to change skin color. Let's suppose my character stayed in his parent's basement all his time before the Cataclysm, and after surviving to summer and autumn, going outside, I want to change his skin into a darker tone. For changing hairstyles, the possible roleplay reasons are more self evident. As for the eyes, maybe I'd like to change their color after messing with mutations, while still keeping it immersive.

Of course, all of those examples could be automated and become "immersive features", and I would love to see it, but it's unlikely to happen anytime soon, as we have not seen any implementation of hair growth yet, or tanning which I presume would have to update a "tanning" value every single turn you are (or aren't) exposed to the sun, which sounds like it would affect performance. Letting the player change the cosmetic sprites can be immersive! Although I also might just want to use it as a second chance to customize my character (which may be good, if I start with a ridiculous-looking random character that I want to give an actual dedicated look for).

**Comment by Theawesomeboophis** [comment]
But sprites are cosmetic and dependent on tileset, there is no actual change to your character's appearance as far as the game is concerned, it's just changing your pronouns.

**Comment by I-am-Erk** [comment]
Additionally, unless you're wandering around naked, the game makes no distinction on if you're presenting as male because you look male or because you're wearing a chest binding. And this still doesn't represent any change of biology, it's just iconography. Consider it "doing a Mulan".

I'm not strongly opposed to editing other features in this way, but I actually would like to *avoid* furthering the misconception that changing your expressed gender has to do with changing immutable parts of your body.

**Comment by DrPariah** [comment]
This is something I was wondering about myself. It's literally cosmetics, it has no game play effect, and if you can extend the functionality to other cosmetics it would allow you to apply scars, tattoos, and other visual cues. Got your legs broken and recovered from it? Add a scar to remind you of your mauling and near escape. Got a faction camp? Get a tattoo to signify your new faction. Want to make your survivor grizzled a year in, or just not shave? Add a beard. I, for one, would adore this functionality and I don't think I'm the only one.
(and as pointed out repeatedly, there's zero mechanical impact for this, nor does there need to be, so there's no need to lock it behind items or NPCs who may or may not be available)

**Comment by I-am-Erk** [comment]
Like I said, I'm not vehemently opposed to it. I think it might still be best suited to an action in the game, personally - for example, an interaction using a mirror could allow you to customize any part of your appearance, the haircut kit can (and should) be allowed to customize just your hair, etc. This is more in line with how car parts appearances are changed too.

**Comment by Theawesomeboophis** [comment]
We could possibly keep barbers for when the character doesn't have a mirror nearby, I personally think the mirror thing is a great idea, maybe a check for if you have appropriate items along with the mirror, like if you don't have trimmers you can't cut your hair or something.

**Comment by Theawesomeboophis** [comment]
I think the mirror thing is a great idea, maybe we can keep the barbers for when the character doesn't have a mirror handy.

**Comment by Theawesomeboophis** [comment]
I also think scars should be implemented when a proper wound system is, as I think your character retaining wounds from previous encounters is a neat and fun idea.

**Comment by Saicchi** [comment]
> for example, an interaction using a mirror 

Oh. I **really** like this mirror idea.

**Comment by Theawesomeboophis** [comment]
Maybe a check for if you actually have the right equipment, example: if you don't have scissors it could give you a message like " you're not edward scissorhands".

**Comment by I-am-Erk** [comment]
We could keep mirrors for skin and eyes and other details, and have hair be barbers or the haircut kit.

**Comment by acepleiades** [comment]
> In fact now that that has occurred to me, could we please do it that way?

You're the higher-up so your word is final.

**Comment by DrPariah** [comment]
Seems acceptable. Those are easy to find items, so it's not making the player jump through too many hoops.

**Comment by Saicchi** [comment]
I've added a new field, called `vanity` which is a bool. Only fields that have this vanity set to `true` can be changed at any time.
Eyes and skin colour can be changed by using a hand mirror ( item ) or a standing mirror ( furniture ).
Hair and facial hair can be changed by getting a haircut using the `makeshift haircut kit` or the `electric hair trimmer`.
I'll update the PR body shortly to match.

**Comment by Theawesomeboophis** [comment]
Awesome job.

File: src/mutation.cpp [hunk 1]
```diff
@@ -1860,6 +1861,79 @@ bool contains_trait( std::vector<string_id<mutation_branch>> traits, const trait
...
+            break;
+        case customize_appearance::HAIR_F:
+            amenu.text = _( "Choose a new facial hairstyle" );
+            traits = get_mutations_in_type( STATIC( "facial_hair" ) );
+            end_message = _( "Surviving the end with style." );
+            break;
+        case customize_appearance::SKIN:
+            amenu.text = _( "Choose a new skin colour" );
+            traits = get_mutations_in_type( STATIC( "skin_tone" ) );
+            end_message = _( "You activate your secret chameleon powers." );
```
Comment (I-am-Erk) [inline src/mutation.cpp:1909]:
```suggestion
            end_message = _( "Life in the cataclysm seems to have changed you." );
```

**Comment by I-am-Erk** [comment]
Ok. I'm happy with this and will merge it once the tests are done, unless someone raises concerns.

**Comment by I-am-Erk** [comment]
Throwing some errors:

> src/mutation.cpp: In member function 'void Character::customize_appearance(customize_appearance)':
> Error: src/mutation.cpp:1891:14: error: 'customize_appearance' is not a class, namespace, or enumeration

**Pull request merged**